### DISCUSSION TITLE
Strangities discoverd while following the memory example

### DISCUSSION THREAD

**Discussion by Zykino** [slint-ui/slint#2717] [Q&A] [answered]
I wanted to report back some strange behavior from the example. I’m on pop-os (ubuntu based) with rust v1.69.0, and slint 1.0.

First the questions I have on slint. (I’m kind of used to QML so the object/JSON notation is like home.)
1. The screenshoths/gif/previews are all centered but on my side I have the tiles locked at the top left of the screen. I don’t see how to center them: the `GridLayout` does not accept `for` or `if`, combining `HorizontalLayout` and `VerticalLayout` I cannot cut the loop in multiples with ranges `0..4` …
2. How can I make the curtain have a border but not in the center? When I try I get 2 ovals on top of the tile (so with a yellow border on the tile (or different color background between tile and curtain), I can see the tile when the curtain should be closed. Is there a `border-direction`?
3. In a memory we should not be able to flip each card and then select the combo when found ;)

I also have some suggestions / bugs to report:
1. `tiles.extend_from_within(..);` works. If it is specialized I suppose its "better" than `tiles.extend(tiles.clone());`. But I’m still new to rust so I’m happy to understand if they are similar or one is really better than the other.
2. It perturbed me a lot to not be able to have comments on the same line as code. Both `//` and `/* */` cryptic give compilation errors.
4. `in-out property <[TileData]> memory_tiles: [ /* ... */ ]` the `-out` part is not necessary. While leaving out the `in` gives correctly the error `error[E0624]: method `set_memory_tiles` is private`. I was surprised to see the code compile while seeing I forgot the `-out`. It also feels strange to have the possibility of mirror between in and out, but having one supersede the other in the end.

**Comment by tronical** [comment, marked as answer]
Hi! Thanks for your feedback (and thanks for your patience ;)

1. Until #358 is implemented, I think you'd have to do what we do in the slide puzzle by computing the equidistant grid placement manually, like [here](https://github.com/slint-ui/slint/blob/dd04e044d0c1c585545d41c62f506bf6a325bbfb/examples/slide_puzzle/slide_puzzle.slint#L274). A resizable memory game would actually be a nice addition :)
2. Can you elaborate what you mean with border but not in the center? Something like #157?

Regarding the `tiles.extend_from_within` use: You're right, that could be used instead. I'm not sure it makes a difference performance wise. I think `extend` is a little simpler to understand though. But clearly a matter of preference :)

Regarding the `//` and `/* */`, I'm not sure I understand. I seem to be able to mix them, but maybe I'm missing something. Can you describe an example where this fails?

For the `in-out` property: I see that we use just `in` but not `in-out`. Is that somewhere wrong in the tutorial? I agree, it should be just `in`.

**Reply by Zykino** [reply]
I think yes, having individual border radius should help me have a better curtain here, without rounding at the center (the intrudoctury gif have no center rounding https://slint-ui.com/blog/memory-game-tutorial/memory_clip.mp4).
![image](https://github.com/slint-ui/slint/assets/3809938/dd593e94-17f4-48b1-876c-3b1b2e45971a)

I don’t remember for the comments, I had issues having them on the same line as code, but I don’t seems to be able to reproduce it right now. Maybe it was on some specific lines.

`in-out` presented [here](https://slint-ui.com/releases/1.0.2/docs/tutorial/rust/game_logic_in_rust) while only `in property <[TileData]> memory_tiles: [` still works.
